hail/src/main/scala/is/hail/expr/types/physical/PNDArray.scala

```diff
@@ -29,4 +30,76 @@ final case class PNDArray(elementType: PType, nDims: Int, override val required:
   override def unsafeOrdering(): UnsafeOrdering = representation.unsafeOrdering()
 
   override def fundamentalType: PType = representation.fundamentalType
+
+  def numElements(shape: Code[Long], mb: MethodBuilder): Code[Long] = {
+    def getShapeAtIdx(idx: Int): Code[Long] = Region.loadLong(this.representation.fieldType("shape").asInstanceOf[PTuple]
+      .loadField(shape, idx))
+
+    if (nDims == 0) {
+      0L
+    }
+    else {
+      Array.range(0, nDims).foldLeft(Code(1L)) { (prod, idx) => prod * getShapeAtIdx(idx) }
```

Comment:
you can call the `const` function in the asm4s package for the Long -> Code[Long] coercion and you can also explicitly specify the type parameters to the `foldLeft` function.

Reply:
Thanks, forgot about these options